### ISSUE TITLE
Columnar metapage is read as all zeros under high write concurrency

### ISSUE THREAD

**Issue by onurctirtir** [citusdata/citus#5333]
(Only applies to 10.2, still investigating)

```sql
CREATE TABLE test (a int) USING COLUMNAR;
```

```
$ cat test.sql
INSERT INTO test SELECT i FROM generate_series ((10000*random())::int, (10000*random())::int) i;
### end of test.sql ###

$ pgbench -c 32 -j 16 -T 600 -P 10 -r --host=localhost --port=10700 -f test.sql
...
ERROR: attempted to access relation "test", which uses an older columnar format
DETAIL: Columnar format version 2.0 is required, "test" has version 0.0.
...
```

**Comment by jeff-davis** [comment]
In `WriteToBlock()`, if `clear == true`, then it will clear the page before writing the new one, rather than just adding data to the page. That means any concurrent connection that is holding only a pin will be able to see the all-zero state between the `InitPage()` and the `memcpy_s()`.

To remove data from a page, or do anything else that might interfere with someone who has a pin on the page, the standard way to do that is with `LockBufferForCleanup()`, which takes a `BUFFER_LOCK_EXCLUSIVE` and waits for the pin count to drop to just itself. But `WriteToBlock()` uses a `BUFFER_LOCK_EXCLUSIVE` only.

We have a few options:

1. Take a `BUFFER_LOCK_SHARED` instead of just a pin, as in #5338 when reading, so that it conflicts with `BUFFER_LOCK_EXCLUSIVE`.
2. Call `LockBufferForCleanup()` when modifying the metapage in a way that interferes with pin holders.
3. Avoid clearing the metapage for simple operations like reserving row and stripe numbers (would require an atomic 64-bit update or something).

The first approach is the simplest, and it doesn't seem like it would cause extra contention. It does go through the locking code path, so might be a bit heavyweight for the typical read workload, but perhaps a quick measurement will tell us if there's a problem there (unlikely to have a major effect compared to decompression).

It feels slightly off to take a share lock for a simple read of the data pages, though, as it's not typically how things are done in postgres. I think what this shows is that the metapage is really different from the data pages, and I tried too hard to combine those code paths. It might make more sense to just have separate read and write paths for the metapage vs the data pages.

In other words:

* `ReadFromBlock()`/`WriteFromBlock()` would assert that they are not touching the metapage
* `ReadFromBlock()` would only require a Pin (unchanged from today)
* `WriteFromBlock()` would take an exclusive lock, and only append data to a page (after `pd_lower`)
* `ColumnarMetapageRead()`/`ColumnarOverwriteMetapage()` would assert that they only touch the metapage
* `ColumnarMetapageRead()` would take a share lock and read the whole page (the metapage would never be read with just a pin)
* `ColumnarOverwriteMetapage()` would take an exclusive lock, and clear and write the whole page
